src/main/java/fi/vm/yti/datamodel/api/v2/service/SchemaService.java

```diff
@@ -1,161 +1,264 @@
 			nodeShapeResource.addProperty(DCTerms.description, node.get("description").asText());
-		nodeShapeResource.addProperty(model.getProperty(SHACL.name.getURI()), propID);
-		if(node == null || node.get("properties") == null) {
+		if (node == null || node.get("properties") == null) 
 			return;
-		}
-		
-		/* Iterate over properties
-		 * If a property is an array or object – add and recursively iterate over them
-		 * If a property is a datatype / literal – it's just added.
+		if (node.has("additionalProperties"))
+			nodeShapeResource.addProperty(SH.closed, model.createTypedLiteral(!node.get("additionalProperties").asBoolean()));
+		/*
+		 * Iterate over properties If a property is an array or object – add and
+		 * recursively iterate over them. If a property is a datatype or literal – it's just added.
 		 */
+
 		Iterator<Entry<String, JsonNode>> propertiesIterator = node.get("properties").fields();
-		while(propertiesIterator.hasNext()) {
+		while (propertiesIterator.hasNext()) {
 			Entry<String, JsonNode> entry = propertiesIterator.next();
-			if(entry.getKey().startsWith("_") || entry.getKey().startsWith("$"))
+			
+			if (entry.getKey().startsWith("_") || entry.getKey().startsWith("$"))
 				continue;
-			//System.out.println(entry.getValue());
-			if(entry.getValue().get("type") != null && entry.getValue().get("type").asText().equals("object")) {
-				// add object property 
-				Resource propertyShape = addObjectProperty(entry.getKey(), entry.getValue(), model, schemaPID, schemaPID + "#" + propID + "/" + entry.getKey());
-				nodeShapeResource.addProperty(model.getProperty(SHACL.property.getURI()), propertyShape);
-				handleObject(propID + "/" + entry.getKey(), entry.getValue(), schemaPID, model);
-			}
-			else if(entry.getValue().get("type") != null && entry.getValue().get("type").asText().equals("array") && entry.getValue().get("items").has("type") && entry.getValue().get("items").get("type").asText().equals("object")) {
+			if (entry.getValue().get("type") == null) 
+				throw new RuntimeException("One of the entries is missing 'type' property");
+
+			if (entry.getValue().get("type").asText().equals("object")) {
+				Resource propertyShape = addObjectProperty(propIDCapitalised + "/" + entry.getKey(), entry.getValue(), model, schemaPID,
+						schemaPID + "#" + propIDCapitalised + "/" + entry.getKey());
+				nodeShapeResource.addProperty(model.getProperty(SH.property.getURI()), propertyShape);
 				
-				Resource propertyShape = addObjectProperty(entry.getKey(), entry.getValue(), model, schemaPID, schemaPID + "#" + propID + "/" + entry.getKey());
-				nodeShapeResource.addProperty(model.getProperty(SHACL.property.getURI()), propertyShape);	
-				handleObject(propID + "/" + entry.getKey(), entry.getValue().get("items"), schemaPID, model);	
+				handleObject(propIDCapitalised + "/" + entry.getKey(), entry.getValue(), schemaPID, model);
+
+			} else if (entry.getValue().get("type").asText().equals("array")
```

Comment:
This if/elseif/else construct is a bit hard to read. Suggestions to consider:

- Move the next else if inside this one, so that the "array" comparison is done only once.
- Add helper methods for long property paths. For example  entry.getValue().get("items").get("type").asText().equals("object") could be replaced with hasObjectItem(). And the whole condition could become something like: if(isArray(entry) and hasObjectItems(entry).